esphome/components/wts01/wts01.cpp

```diff
@@ -0,0 +1,100 @@
+#include "wts01.h"
+#include "esphome/core/log.h"
+#include <cmath>
+
+namespace esphome {
+namespace wts01 {
+
+static const char *const TAG = "wts01";
+static const uint8_t PACKET_SIZE = 9;
+static const uint8_t HEADER_1 = 0x55;
+static const uint8_t HEADER_2 = 0x01;
+static const uint8_t HEADER_3 = 0x01;
+static const uint8_t HEADER_4 = 0x04;
+
+void WTS01Sensor::loop() {
+  // Process one character at a time received from the sensor
+  if (available()) {
+    uint8_t c;
+    if (read_byte(&c)) {
+      this->handle_char_(c);
+    }
+  }
+}
+
+void WTS01Sensor::dump_config() {
+  ESP_LOGCONFIG(TAG, "WTS01 Temperature Sensor:");
+  LOG_SENSOR("  ", "Temperature", this);
+  this->check_uart_settings(9600);
+}
+
+void WTS01Sensor::handle_char_(uint8_t c) {
+  // Reset buffer if we're at the start of a new packet
+  if (c == HEADER_1 && this->buffer_pos_ == 0) {
+    this->buffer_[this->buffer_pos_++] = c;
+    return;
+  }
+
```

Comment:
You need to reset the state machine/buffer counter if the first byte doesn't match, not just continue if it does match.

```suggestion
  // State machine for processing the header. Reset if something doesn't match.
  if (this->buffer_pos_ == 0 && c != HEADER_1) {
    this->buffer_pos_ = 0;
    return;
  }

```